rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,337 @@
+operators where execution yields a new schema.
+
+##### Filter
+
+| Operator | Non-Blocking | Description
+| -------- |:------------:| ------------------------------------------
+| `head`   | ✅           | Retrieve first N records
+| `tail`   | ❌           | Retrieve last N records
+| `where`  | ✅           | Filter with expression
+| `sort`   | ❌           | Sort according to field
+| `uniq`   | ✅           | Produce deduplicated output
+
+##### Reshape
+
+| Operator    | Non-Blocking | Description
+| ----------- |:------------:| ------------------------------------------
+| `put`       | ✅           | Select a set of columns (projection)
+| `drop`      | ❌           | Remove a set of columns (projection)
+| `replace`   | ✅           | Replaces field described by extractors with values
+| `extend`    | ✅           | Adds new fields with (initially fixed) values
+| `rename`    | ✅           | Renames schema meta data (type & fields)
+| `summarize` | ❌/✅[^1]    | Aggregate group and compute a summary function
+
+[^1]: Depending on the aggregate function.
+
+### Pipeline Execution
+
+Let's take a step back and just assume that we add a new `exec` command that
+just executes a pipeline.
+
+Then we can model the ingestion as follows: load data via stdin, push into a
+parser, then send off to a remote VAST node. For example, we would rewrite
+`vast import zeek` as:
+
+```bash
+vast exec 'from - | read zeek | push vast://1.2.3.4'
+```
+
+Likewise, `vast export json EXPRESSION` would become:
+
+```bash
+vast exec 'pull vast://1.2.3.4 |
+           where EXPRESSION |
+           write json |
+           to -'
+```
+
+The beauty is that VAST nodes are now longer required. Pipelines provide value
+locally as well:
+
+```bash
+vast exec 'from - |
+           where EXPRESSION |
+           write json |
+           to -'
+```
+
+What would the role of the node be then? Basically a **pipeline manager** with
+persistent state. A lot of things wouldn't change, e.g., starting a node:
+
+```bash
+vast -e 1.2.3.4 start
+```
+
+There are of course convenience defaults we can use, like assuming `from -` and
+`to -` being the default. This would make the above look like this:
+
+```bash
+vast exec 'where EXPRESSION |
+           write json
+```
+
+Side effect: we would get our envisioned convert operator almost for free:
+
+```bash
+vast exec 'read zeek | print csv'
+# Obviates the need for:
+vast exec 'convert zeek csv'
+```
+
+### Node Interaction
+
+When considering pipelines in conjunction with VAST nodes, we may want to
+improve the UX of working with pipelines. In the new model, VAST would be
+an `Arrow` source or sink, making the primary interaction on the data plane
+through `push` and `pull` operators.
+
+Making `push` and `pull` top-level commands would streamline UX. For example,
+consider this `vast export` pipeline:
+
+```bash
+# exec
+vast exec 'pull vast://1.2.3.4 |
+           where EXPRESSION |
```

Comment:
Shouldn't it be `pull URI EXPR` instead of `pull URI | EXPR` because you want to execute that query remotely? Unless there is some implied optimization that lifts `where` operators into preceding `pull` and `push` operators.

Reply:
I don't think we want `pull URI EXPR`. We want to orthogonalize loading from filtering. It's the job of an optimizer to do predicate pushdown.

A query is much more than `where EXPR`, e.g., also includes `summarize` and other operators.

@rdettai pointed out the overarching issue we need to resolve: the location where a pipeline runs is not yet unambiguous in the current proposal.

Reply:
I think it would be the job of the optimizer to transform from `pull URI | where EXPR` to `pull URI EXPR`, but that doesn't mean we shouldn't be able to write the optimized case already.

Predicate pushdown is a hard problem to solve, and if you make it a requirement for this proposal you'll just push it further down the road.

Reply:
I guess this is also an implication of what I raised [here](https://github.com/tenzir/vast/pull/2511/files#r954679555).
- `pull URI EXPR` has the benefit of letting the user tune where he wants things to be executed. It makes the `pull` primitive more complex, as you would need to be able to run en entire pipeline there. Having `pull` within the pipeline would make the syntax a bit clumsy (`vast exec 'pull \'filter EXPR\' | write json`), having it as a dedicated command would read better (`vast pull 'filter EXPR' | write json`)
- `pull URI | filter EXPR` requires predicate pushdown and lifting the filter to the remote. The query becomes automatically optimized. A user wanting to force the system to resolve the filter locally could in theory still doing so by breaking the pipeline `vast exect 'pull URI' | vast exec 'filter EXPR'`  as the UNIX pipe will require the stream to be local. This syntax is flatter and a bit more "magic", but requires a fair amount of query planning logic to be implemented.

Reply:
> but that doesn't mean we shouldn't be able to write the optimized case already.

With most query languages, you can't (e.g SQL). Not saying that's a good thing, but it's definitively a common practice.

> Predicate pushdown is a hard problem to solve, and if you make it a requirement for this proposal you'll just push it further down the road

From my experience on datafusion, query optimization (and predicate pushdown as a part of it) is actually something you can do very incrementally, as soon as you have a good representation of the query plan. We could maybe even argue that forcing ourselves to frame a good query plan representation upfront is something crucial, as it might be hard to refactor afterward.

Reply:
(Replacing `filter` with `where` mentally in my head.)

Providing a filter predicate to `pull` feels wrong at first, because `push` and `pull` should have a single function: send or receive data to a VAST node. For filtering, you'd use `where`. If we start allowing filtering in multiple commands, we're losing clarity.

Reply:
Per https://github.com/tenzir/vast/pull/2511#discussion_r956371796, we should perhaps decide also between live, historical, and unified querying. Perhaps `pull` should accommodate for that?

Reply:
Resolving this thread, as `pull` is no longer in here. To be continued in #2577.